text/0000-feature-metadata-doc-cfg.md

```diff
@@ -0,0 +1,334 @@
+[features.qux]
+requires = ["bar", "baz"]
+doc = """
+# qux
+
+This could be a longer description of this feature
+"""
+```
+
+This RFC will also allow a `[tools.rustdoc]` table where existing configuration
+can be specified:
+
+```toml
+# Cargo.toml
+
+[tools.rustdoc]
+html-logo-url = "https://example.com/logo.jpg"
+issue-tracker-base-url = "https://github.com/rust-lang/rust/issues/"
+```
+
+For projects that do not use Cargo or want separate configuration, these options
+can also be specified in a `rustdoc.toml` file using an identical schema:
+
+```toml
+# rustdoc.toml containing the same information as above
+# the only difference is that [tools.rustdoc] has become top level
+html-logo-url = "https://example.com/logo.jpg"
+issue-tracker-base-url = "https://github.com/rust-lang/rust/issues/"
+
+[features]
+foo = []
+bar = { requires = ["foo"], doc = "simple docstring here" }
+# ...
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+At a high level, the following changes will be necessary:
+
+1. Cargo will change its parsing to accept the new format for `[features]`
+2. `rustdoc` will gain two optional arguments: `--config-file` (for specifying
+   `rustdoc.toml`-style files), and `--config-json` (for specifying the same
+   information via JSON text, or a path to a JSON file). These arguments can be
+   specified more than once
+3. Cargo will naively serialize some information from `Cargo.toml` to pass to
+   `rustdoc`, and make `rustdoc` aware of any `rustdoc.toml` files.
+4. `rustdoc` will parse each of the `--config-*` arguments to create its
+   internal configuration.
+
+This is described in more detail in the following sections.
+
+## Changes to Cargo
+
+Cargo will need to parse the new format for `[features]`. For its non-rustdoc
+use, it can discard all new information.
+
+The `cargo doc` invocation will need to do two new things:
+
+1. Reserialize the `[features]` and `[tools.rustdoc]` tables from any
+   `Cargo.toml` file to JSON. This can be naive, i.e., Cargo does not need to
+   validate the contained information in any way.
+2. Pass this information as a string via `--config-json`. If string length
+   exceeds a limit (e.g., 2000 characters), write this configuration instead to
+   a temporary build JSON file. (to avoid maximum argument length restrictions
+   and keep verbose output concise).
+3. Find files named `rustdoc.toml` and pass their paths to `rustdoc` using
+   `--config-toml`; no parsing of these files is necessary.
+
+Cargo should use the following precedence (later items take priority over
+earlier items):
+
+1. Workspace `Cargo.toml`
+2. Workspace root `rustdoc.toml`
+3. Crate `Cargo.toml`
+4. Crate root `rustdoc.toml`
+
+`rustdoc` will be in charge of handling configuration merging. This should
+create an intuitive layering of global options and overrides while keeping
+`rustdoc` and `Cargo` reasonably separate.
```

Comment:
Sounds like this design has implicit workspace inheritance?  Seems like we should start off with explicit workspace inheritance to be consistent with the work already done.  See also rust-lang/cargo#8415

Reply:
This is referring to e.g. `version.workspace = true` in https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table, correct?

The reason I opted for the way it is written is that I think it does a good job of keeping build system and tool decoupled. The build system only needs to inform the tool that there is a precedence/inheritance in configuration: the tool then goes about how to merge the configuration. I think this allows for more generic tooling (e.g. with https://github.com/rust-lang/rfcs/pull/3416#discussion_r1169139288)

(But I am happy to remove or rework this inheritance section)

Reply:
I described this more in https://github.com/rust-lang/rfcs/pull/3421, we can probably have some more discussion there